Service input plugin for OpenConfig streaming telemetry

ajhai opened the pull request:
### Required for all PRs:

- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

This pull request adds a service input plugin that allows telegraf to collect streaming telemetry over gRPC from network devices that support [OpenConfig](http://openconfig.net/) streaming telemetry. Since OpenConfig data is a list of key value pairs, fields and tags vary depending on the source of data.

ajhai commented:
Hey @danielnelson , I fixed the merge conflicts and failing tests. Could you please take a look?

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+	sync.Mutex
+
+	// keep the accumulator internally:
+	acc            telegraf.Accumulator
```
I don't think you need to save this to the struct, since you pass it to the goroutines from the Start functions.

ajhai replied:
You're right. Don't need it. Changed it accordingly

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/README.md:
```diff
@@ -0,0 +1,46 @@
+  server = ["localhost:1883"]
+
+  ## Frequency to get data in milliseconds
+  sampleFrequency = 2000
```
Make this an `internal.Duration` so it can be specified like: `"2000ms"`

ajhai replied:
Using internal.Duration for sampleRrequency/reportingRate

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+	c := telemetry.NewOpenConfigTelemetryClient(m.grpcClientConn)
+
+	for _, sensor := range m.Sensors {
+		go func(sensor string, acc telegraf.Accumulator) {
```
Use a `sync.WaitGroup` to ensure these goroutines exit on Stop, this is important for reloading the Telegraf config.

ajhai replied:
Done.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+func (m *OpenConfigTelemetry) Start(acc telegraf.Accumulator) error {
+	log.Print("I! Started JTI OpenConfig Telemetry plugin\n")
+
+	m.Lock()
```
You may be able to remove this lock, Telegraf guarantees it will serialize calls to Start(), Stop(), etc.

ajhai replied:
Removed it.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/README.md:
```diff
@@ -0,0 +1,46 @@
+  sensors = [
+   "/interfaces/",
+   "collection /components/ /lldp",
+  ]
```
Add example of custom reporting rate, and explain how the collection works.

ajhai replied:
Done!

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+			stream, err := c.TelemetrySubscribe(context.Background(),
+				&telemetry.SubscriptionRequest{PathList: pathlist})
+			if err != nil {
+				log.Fatalf("E! Could not subscribe: %v", err)
```
Use `acc.AddError`, also should we return from the goroutine here if we can't subscribe.

ajhai replied:
Sure.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+			reportingRate = uint64(m.SampleFrequency)
+
+			if len(spathSplit) > 1 {
+				for i, path := range spathSplit {
```
This is more complex than it needs to be, I don't think you should use a loop here.  Try assigning the components of the sensors option to variables and then take action based on their values.

ajhai replied:
Thanks. Don't know what I was thinking before. Rewrote this.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry_test.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+func TestMain(m *testing.M) {
+	lis, err := net.Listen("tcp", "127.0.0.1:50051")
```
Listen on a random port by using port 0.

ajhai replied:
Done!

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+
+				dgroups := []DataGroup{}
+
+				for _, v := range r.Kv {
```
Can you provide some sample data from an actual device so that I can better evaluate the data schema?

ajhai replied:
system_id:jtidevice
component_id:65535
path:sensor_1000:/interfaces/interface/state/oper-status/:/interfaces/interface/state/oper-status/:mib2d
sequence_number:0
timestamp:1513886669462
kv {
  key:__timestamp__
  uint_value:1513886671466
}
kv {
  key:__junos_re_stream_creation_timestamp__
  uint_value:1513886669430
}
kv {
  key:__junos_re_payload_get_timestamp__
  uint_value:1513886669430
}
kv {
  key:__prefix__
  str_value:/interfaces/interface[name='lsi']/
}
kv {
  key:state/oper-status
  str_value:UP
}

danielnelson commented:
Another question I have is how specific this is to Juniper Networks devices vs OpenConfig in general.  Could this be used with any device supporting OpenConfig?

Also, do you have any advice for testing against a real server?  Is it best to buy a actual hardware device or maybe there is a software implementation or emulator?

ajhai commented:
@danielnelson thanks for reviewing the changes. I'll address your comments and update the pull request after testing my changes. For your other questions, this plugin is specific to Juniper's implementation of OpenConfig telemetry in terms of how we do client authentication, gRPC APIs that we expose etc. OpenConfig uses yang to define models and these APIs. Since there is no one way to translate from yang to proto, service and API naming may differ in OpenConfig telemetry implementation. Otherwise it should work fine as long as the APIs match.

Juniper ships a VM that can be used to test this. https://www.juniper.net/us/en/products-services/routing/mx-series/vmx/. We use this plugin internally with real devices in our lab.

ajhai commented on plugins/inputs/jti_openconfig_telemetry/oc/oc.proto:
```diff
@@ -0,0 +1,289 @@
+//
```
These files are available to download from juniper.net as a package at https://webdownload.juniper.net/swdl/dl/secure/site/1/record/72272.html . Couldn't figure out a better way to bring them in than this.

ajhai commented on Godeps:
```diff
@@ -76,10 +76,12 @@ github.com/wvanbergen/kafka bc265fedb9ff5b5c5d3c0fdcef4a819b3523d3ee
 github.com/wvanbergen/kazoo-go 968957352185472eacb69215fa3dbfcfdbac1096
 github.com/yuin/gopher-lua 66c871e454fcf10251c61bf8eff02d0978cae75a
 github.com/zensqlmonitor/go-mssqldb ffe5510c6fa5e15e6d983210ab501c815b56b363
+google.golang.org/grpc 54222d50cc54928f54ebb47ad55797858e0a0268
```
Fixed this!

ajhai commented on plugins/inputs/jti_openconfig_telemetry/README.md:
```diff
@@ -0,0 +1,46 @@
+  server = ["localhost:1883"]
+
+  ## Frequency to get data in milliseconds
+  sampleFrequency = 2000
```
Done!

danielnelson replied:
Don't forget to update the sample in this README, I usually just run something like `telegraf -usage jti_openconfig_telemetry` and copy paste it in.

ajhai commented on plugins/inputs/jti_openconfig_telemetry/README.md:
```diff
@@ -0,0 +1,46 @@
+  certFile = "/path/to/x509_cert_file"
+
+  ## Option to debug incoming protobuf encoded data
+  debug = true
```
Done!

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+	Server          string
+	Username        string
+	Password        string
+	ClientId        string
```
Done!

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+
+  ## x509 Certificate to use with TLS connection. If it is not provided, an insecure 
+  ## channel will be opened with server
+  certFile = "/path/to/x509_cert_file"
```
Done!

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+
+// Takes in XML path with predicates and returns list of tags+values along with a final XML path without predicates
+func spitTagsNPath(xmlpath string) (string, map[string]string) {
+	re := regexp.MustCompile("\\/([^\\/]*)\\[([A-Za-z0-9\\-\\/]*)\\=([^\\[]*)\\]")
```
I'm extracting some data out of XML path here. I couldn't find a way to do with encoding/xml. Simplified this function.

danielnelson replied:
We have a dependecy on the gjson library, perhaps you can use this?

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+	return xmlpath, tags
+}
+
+func (m *OpenConfigTelemetry) Start(acc telegraf.Accumulator) error {
```
Done

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+func (m *OpenConfigTelemetry) Start(acc telegraf.Accumulator) error {
+	log.Print("I! Started JTI OpenConfig Telemetry plugin\n")
```
Changed it to debug

danielnelson replied:
I think you can completely remove this, Telegraf will print out the enabled inputs on startup.

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,324 @@
+	if m.CertFile != "" {
+		creds, err := credentials.NewClientTLSFromFile(m.CertFile, "")
+		if err != nil {
+			log.Fatalf("E! Failed to read certificate: %v", err)
```
Done. Made changes here and other similar places.

ajhai commented:
On Fri, Jan 26, 2018 at 12:54 PM, Daniel Nelson <notifications@github.com>
wrote:

> *@danielnelson* commented on this pull request.
> ------------------------------
>
> In plugins/inputs/jti_openconfig_telemetry/README.md
> <https://github.com/influxdata/telegraf/pull/2292#discussion_r164219270>:
>
> > @@ -0,0 +1,46 @@
> +# JTI OpenConfig Telemetry Input Plugin
> +
> +This plugin reads Juniper Networks implementation of OpenConfig telemetry data from listed sensors using Junos Telemetry Interface. Refer to
> +[openconfig.net](http://openconfig.net/) for more details about OpenConfig and [Junos Telemetry Interface (JTI)](https://www.juniper.net/documentation/en_US/junos/topics/concept/junos-telemetry-interface-oveview.html).
> +
> +### Configuration:
> +
> +```toml
> +# Subscribe and receive OpenConfig Telemetry data using JTI
> +[[inputs.jti_openconfig_telemetry]]
> +  server = ["localhost:1883"]
> +
> +  ## Frequency to get data in milliseconds
> +  sampleFrequency = 2000
>
> Don't forget to update the sample in this README, I usually just run
> something like telegraf -usage jti_openconfig_telemetry and copy paste it
> in.
>

Done. Sorry missed this earlier.


> ------------------------------
>
> In plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go
> <https://github.com/influxdata/telegraf/pull/2292#discussion_r164219626>:
>
> > +	m.parser = parser
> +}
> +
> +func (m *OpenConfigTelemetry) Gather(acc telegraf.Accumulator) error {
> +	return nil
> +}
> +
> +func (m *OpenConfigTelemetry) Stop() {
> +	m.Lock()
> +	defer m.Unlock()
> +	m.grpcClientConn.Close()
> +}
> +
> +// Takes in XML path with predicates and returns list of tags+values along with a final XML path without predicates
> +func spitTagsNPath(xmlpath string) (string, map[string]string) {
> +	re := regexp.MustCompile("\\/([^\\/]*)\\[([A-Za-z0-9\\-\\/]*)\\=([^\\[]*)\\]")
>
> We have a dependecy on the gjson library, perhaps you can use this?
>
>
I just checked gjson. Not sure if I can use that here. This function takes
in a XML Path like /sensor/data/key1['tag=tagVal']/key2['tag=tagVal']/point
and extracts /sensor/data/key/ and returns (/sensor/data/key1/@tag:tagVal,
/sensor/data/key1/key2/@tag:tagVal) and /sensor/data/key1/key2/point.
Please let me know if I'm missing something.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,353 @@
+				}
+				if err != nil {
+					acc.AddError(fmt.Errorf("E! Failed to read: %v", err))
+					return
```
If an error occurs we need to make sure that we can recover.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,353 @@
+			for {
+				r, err := stream.Recv()
+				if err == io.EOF {
+					break
```
What is the right course of action here?  Do we just ignore this sensor from this point on or do we need to reconnect or something else?

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,353 @@
+				&telemetry.SubscriptionRequest{PathList: pathlist})
+			if err != nil {
+				acc.AddError(fmt.Errorf("E! Could not subscribe: %v", err))
+				return
```
Do we need to reconnect in the future, or can we assume if we couldn't subscribe we never will be able to?

ajhai commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,353 @@
+// is given input, this function will emit /events/event/attributes/value as xmlpath and
+// { /events/event/@id=2, /events/event/attributes/@key='message' } as tags
+func spitTagsNPath(xmlpath string) (string, map[string]string) {
+	re := regexp.MustCompile("\\/([^\\/]*)\\[([A-Za-z0-9\\-\\/]*\\=[^\\[]*)\\]")
```
We are not processing XML documents. Instead simply picking some data from a XPATH string. I looked at existing modules and couldn't find any functionality that extracts this data from given string. All we need is the above regexp to extract the data. xquery can be used to apply given XPATH built using xpath-go on a XML document to extract data. I looked at antchfx/xpath to see if there is a way to easily pull data off from built xpath object. Couldn't find any straightforward way. Please let me know what you think.

danielnelson replied:
We can leave it here for now, though I will be so grumpy if I have to debug it. :)  Can you move it into a package level global though?

ajhai replied:
Thanks. Hopefully we don't have to debug this :-) I have made this global.

danielnelson replied:
Actually I was referring to `re := regexp.MustCompile("\\/([^\\/]*)\\[([A-Za-z0-9\\-\\/]*\\=[^\\[]*)\\]")`, if we make this package level we will only need to compile it once.
```go
vars (
    // come up with a longer name now that its package level
	fooRegex = regexp.MustCompile("\\/([^\\/]*)\\[([A-Za-z0-9\\-\\/]*\\=[^\\[]*)\\]")
)
```

You can leave the function unexported with the lowercase name.

ajhai commented:
@danielnelson I updated code to address your last round of comments. Took care of connection failure fallbacks and added ability to specify list of servers to use same sensor configuration.

danielnelson commented on plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go:
```diff
@@ -0,0 +1,353 @@
+		dgroups = CollectionByKeys(dgroups).Insert(finaltags,
+			map[string]interface{}{"_sequence": r.SequenceNumber})
+		dgroups = CollectionByKeys(dgroups).Insert(finaltags,
+			map[string]interface{}{"_timestamp": r.Timestamp})
```
Do you think we should do this?

ajhai replied:
Timestamp field resolution may not be consistent. Openconfig spec specifies it be in nanoseconds but doesn't mandate it. https://github.com/openconfig/reference/blob/master/rpc/openconfig/openconfig.proto#L64 . There is a chance of some devices sending timestamp in milliseconds and it complicates things. time.Now seems reasonable choice given the possible inconsistencies.

ajhai commented:
@daemonkeeper this plugin handles telemetry data in OpenConfig key value pair format. We did some work on Junos that is going to emit data that used to come out as analyticsd format in OpenConfig format.

ajhai commented:
@danielnelson  Exported SpitTagsNPath and updated PR. Timestamp field resolution may not be consistent. Openconfig spec specifies it be in nanoseconds but doesn't mandate it. https://github.com/openconfig/reference/blob/master/rpc/openconfig/openconfig.proto#L64 . There is a chance of some devices sending timestamp in milliseconds and it complicates things. time.Now seems reasonable choice given the possible inconsistencies. Please let me know if there are any other comments.
